bundle/src/bundle_execution.rs

```diff
@@ -0,0 +1,1184 @@
+use {
+    itertools::izip,
+    log::{debug, trace},
+    solana_ledger::token_balances::collect_token_balances,
+    solana_measure::{measure, measure_us},
+    solana_program_runtime::timings::ExecuteTimings,
+    solana_runtime::{
+        account_overrides::AccountOverrides,
+        accounts::TransactionLoadResult,
+        bank::{
+            Bank, LoadAndExecuteTransactionsOutput, TransactionBalances, TransactionExecutionResult,
+        },
+        transaction_batch::TransactionBatch,
+    },
+    solana_sdk::{
+        account::AccountSharedData,
+        bundle::SanitizedBundle,
+        pubkey::Pubkey,
+        saturating_add_assign,
+        signature::Signature,
+        transaction::{SanitizedTransaction, TransactionError, VersionedTransaction},
+    },
+    solana_transaction_status::{token_balances::TransactionTokenBalances, PreBalanceInfo},
+    std::{
+        cmp::{max, min},
+        time::{Duration, Instant},
+    },
+    thiserror::Error,
+};
+
+#[derive(Clone, Default)]
+pub struct BundleExecutionMetrics {
+    pub num_retries: u64,
+}
+
+/// Contains the results from executing each TransactionBatch with a final result associated with it
+/// Note that if !result.is_ok(), bundle_transaction_results will not contain the output for every transaction.
+pub struct LoadAndExecuteBundleOutput<'a> {
+    bundle_transaction_results: Vec<BundleTransactionsOutput<'a>>,
+    result: LoadAndExecuteBundleResult<()>,
+    metrics: BundleExecutionMetrics,
+}
+
+impl<'a> LoadAndExecuteBundleOutput<'a> {
+    pub fn executed_ok(&self) -> bool {
+        self.result.is_ok()
+    }
+
+    pub fn result(&self) -> LoadAndExecuteBundleResult<()> {
+        self.result.clone()
```

Comment:
need to clone?

Reply:
can't move out of self

Reply:
was thinking return a reference and figure it out upstream but minor